Converting skimmer+categorizer into single processor

lzygala opened the pull request:
I have updated my categorizer to include the option to create a skimmed parquet file. The option is described further in the readme as well as in these slides: https://indico.cern.ch/event/1541370/contributions/6486822/attachments/3056987/5404836/2025_04_25_HbbRun3_Update.pdf

cmantill commented on src/hbb/processors/categorizer.py:
```diff
@@ -0,0 +1,329 @@
+        # only consider first 4 jets to be consistent with old framework
+        jets = jets[:, :4]
+        dphi = abs(jets.delta_phi(candidatejet))
+        selection.add('antiak4btagMediumOppHem', ak.max(jets[dphi > np.pi / 2].btagDeepFlavB, axis=1, mask_identity=False) < 0.3086) 
```
recommend to use ParticleNet instead of deepFlavour for Run-3: https://btv-wiki.docs.cern.ch/ScaleFactors/Run3Summer22/#ak4-b-tagging

cmantill commented on src/hbb/processors/categorizer.py:
```diff
@@ -0,0 +1,329 @@
+        dphi = abs(jets.delta_phi(candidatejet))
+        selection.add('antiak4btagMediumOppHem', ak.max(jets[dphi > np.pi / 2].btagDeepFlavB, axis=1, mask_identity=False) < 0.3086) 
+        ak4_away = jets[dphi > 0.8]
+        selection.add('ak4btagMedium08', ak.max(ak4_away.btagDeepFlavB, axis=1, mask_identity=False) > 0.3086) 
```
the working point (0.3086) will vary depending on the tagger and the year, suggest to put this in a dictionary that could be loaded from the json

cmantill commented on src/hbb/processors/categorizer.py:
```diff
@@ -0,0 +1,329 @@
+        goodmuon = (
+            (events.Muon.pt > 10)
+            & (abs(events.Muon.eta) < 2.4)
+            & (events.Muon.pfRelIso04_all < 0.25)
```
tighten the isolation cut to 0.15 and add IP cuts:
```python
muon_IP = (
            ((abs(muon.eta) < 1.479) & (abs(muon.dz) < 0.1) & (abs(muon.dxy) < 0.05)) |
            ((abs(muon.eta) >= 1.479) & (abs(muon.dz) < 0.2) & (abs(muon.dxy) < 0.1))
        )
```

cmantill commented on src/hbb/processors/categorizer.py:
```diff
@@ -0,0 +1,329 @@
+            (events.Electron.pt > 10)
+            & (abs(events.Electron.eta) < 2.5)
+            & (events.Electron.cutBased >= 2)
+        )
```
recommend mva selection instead of cutbased:

```python
electron_kin = (electron.pt > 15) & (abs(electron.eta) < 2.5)
    electron_iso_ID = (electron.pfRelIso03_all < 0.15) & ( getattr(electron, 'mvaNoIso_WP90') )

    electron_IP = (
        ((abs(electron.eta) < 1.479) & (abs(electron.dz) < 0.1) & (abs(electron.dxy) < 0.05)) |
        ((abs(electron.eta) >= 1.479) & (abs(electron.dz) < 0.2) & (abs(electron.dxy) < 0.1))
    ) 
```

cmantill commented on src/hbb/processors/categorizer.py:
```diff
@@ -0,0 +1,329 @@
+        )
+        nelectrons = ak.sum(goodelectron, axis=1)
+
+        ntaus = ak.sum(
```
I vote that we should get rid of the number of taus cut, it causes more pain in the object review than benefit, we can always add it back if needed

cmantill reviewed (commented):
LGTM, left a few comments for review

lzygala commented:
I resolved all of your comments in this new commit. I moved all of the cms-standard object definitions into objects.py. I also added in a taggers.json to keep track of the tagger working points.

gabihamilton commented on submit.py:
```diff
@@ -0,0 +1,173 @@
+        ship_env=True,
+        memory="10GB",
+        image="coffeateam/coffea-dask-almalinux9:latest",
+        # log_directory="/uscmst1b_scratch/lpc1/3DayLifetime/workerlogs/",
```
workerlogs is owned by lzygala, change the line to:

`log_directory=f"/uscmst1b_scratch/lpc1/3DayLifetime/{os.environ['USER']}",`

gabihamilton commented on README.md:
````diff
@@ -1,43 +1,118 @@
+
+Run the processor for a certain year:
+```bash
+python submit.py 2023
````
Can we maybe add instructions for running in the background and not interactively?

cmantill replied:
What do you mean with this? Maybe you can add instructions to run in the background here?
